cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/StageOperationsValidator.java

```diff
@@ -151,26 +152,55 @@ void validate() {
     // Here outout d from OP1 is redundant, since the d in output schema will always come from OP2.
     // However d will not be in the redundantOutputs map, as we only put the redundant fields if they
     // appear in input of some operation. Such redundancy should cause validation checks to fail.
-    Iterator<Map.Entry<String, List<String>>> iterator = unusedOutputs.entrySet().iterator();
+    Iterator<Map.Entry<String, List<FieldOperation>>> iterator = unusedOutputs.entrySet().iterator();
     while (iterator.hasNext()) {
-      Map.Entry<String, List<String>> next = iterator.next();
+      Map.Entry<String, List<FieldOperation>> next = iterator.next();
       String field = next.getKey();
-      List<String> origins = next.getValue();
+      List<FieldOperation> origins = next.getValue();
 
       if (stageOutputs.contains(field)) {
-        // iterator.remove();
         if (origins.size() > 1) {
           // size will be greater than 1 only if its redundant
           origins.remove(origins.size() - 1);
-          // unusedOutputs.put(field, origins.subList(0, origins.size() - 1));
         } else {
           iterator.remove();
         }
+      } else {
+        // If this field is not in the output schema of the stage, it can still be valid if the operations which
+        // generate it are all input schema fields. For example, a Joiner joins two datasets D1,D2 based on the
+        // joiner key D1.K1, D2.K2, and decides to drop the joiner key in the output schema. The operation
+        // [D1.K1, D2.K2] ->[K1, K2] is a valid even though K1,K2 are not in the output schema. In this case, we need
+        // to check if the input fields of the operation are all from the input schema of the stage.
+        if (origins.size() != 1) {
+          continue;
+        }
+        FieldOperation origin = origins.get(0);
+        switch (origin.getType()) {
+          case READ:
+            // if a read operation, the input is from a source, so no need to check
+            iterator.remove();
+            break;
+          case WRITE:
+            // this will not happen since a write operation will not have output fields, and thus no field will
+            // appear in the unusedOutput
```

Comment:
if this is not expected to happen, it would be good to log a warning if it does happen. If I understand correctly, this will lead to an unused output, which can led to failure to record lineage, right?

Reply:
Yes, if this happens, the FLL will fail to get written since there is unused output

Reply:
That is why I would put a warning here. If somebody ever changes the current logic but does not know that here, we expect a certain behavior, then the code will become incorrect. A warning would at least hint at the problem. Some people would even throw an IllegalStateExn here if this is really not expected.